### ISSUE TITLE
the process wil be so slow when reading batch data from csv files in tensorflow

### ISSUE THREAD

**Issue by liumilan** [tensorflow/tensorflow#2323]
def read_data(filename):

        filename_queue = tf.train.string_input_producer([filename])
        reader = tf.TextLineReader()
        key, value = reader.read(filename_queue)
        record_defaults = [[1.0 for col in range(1)] for row in range(280)]
        record_defaults[279][0] = 1
        a = tf.decode_csv(value, record_defaults=record_defaults)
        data = tf.pack(a[0:278])
        label = a[-1]
        min_after_dequeue = 10000
        capacity = min_after_dequeue + 3 * batch_size
        data_batch, label_batch = tf.train.shuffle_batch([data, label], batch_size=batch_size, capacity=capacity,min_after_dequeue=min_after_dequeue)
        return data_batch, label_batch

def main(argv=None):  

        data,label = read_data(FLAGS.train_file)

        tf.initialize_all_variables()
        with tf.Session() as sess:
                coord = tf.train.Coordinator()
                threads = tf.train.start_queue_runners(coord=coord)
                example, label = sess.run([data, label])
                print (example)

Here is my code above ,why it take 10 minutes to print example data?How can I optimise my code? I have only set batchsize is 2

**Comment by liumilan** [comment]
I have assign min_after_dequeue  to 100,and now it can print the result immediately.
Now I increase batchsize to 20000,and it will take so longer.
In caffe,it only take 198ms to get the data.
Is it other issue that i haven't noticed?
 I0512 11:17:32.966964   787 caffe.cpp:280] ctr_feature_data     forward: 198.561 ms.

**Comment by beopst** [comment]
It is because tf.train.shuffle_batch blocks dequeue operation until the queue size reaches to min_after_dequeue. This is intended for better mixing of elements in queue to shuffle data.

**Comment by liumilan** [comment]
And why when I increase batchsize to 20000,and it will take much more to print .Because of larger data?Can it possible to improve the performance when modify the parameters?
@beopst

**Comment by beopst** [comment]
Yes, it is because of batch_size. In this case, the queue should still block until it has a sufficient amount of elements for the first single batch. 

tf.train.shuffle_batch needs some time only for the first dequeue operation. Once it has elements more than min_after_queue, it consumes much little time since tf.train.shuffle_batch runs in a separate thread. You can see if you call `example, label = sess.run([data, label])` in a loop.

**Comment by mrry** [comment]
If you want training to start sooner, you should reduce the value of `min_after_dequeue`. This parameter is here to ensure good mixing, so you may find that the loss of randomness leads to poorer accuracy, but this depends on your model.

**Comment by liumilan** [comment]
yes,i have called example, label = sess.run([data, label]) in a loop .But it still take much time when batchsize is set 20000。
So reduce batchsize is only one way? it seems not best ways. @beopst

**Comment by liumilan** [comment]
yes,i have called example, label = sess.run([data, label]) in a loop .But it still take much time when batchsize is set 20000。
So reduce batchsize is only one way? it seems not best ways. 
@mrry
